### PR TITLE
Remove metadata from EKS errors to stop log spamming

### PR DESCRIPTION
**Pull request by annablender**
Issue: https://github.com/rancher/rancher/issues/37650 

 
## Problem
 

When a call to `UpdateNodeGroupVersion` fails due to a custom AMI version causing node provisioning to fail in EKS, the failure is expected but lasso spams the EKS logs every second. This happens because the error message that is returned from EKS has metadata with a unique `requestID` in the body. Since the error message is different every time, [this code](https://github.com/rancher/eks-operator/blob/1b8dfe2ea04eee43fd362c6da0e1cef2a09cb600/controller/eks-cluster-config-handler.go#L122) is skipped and the [Status gets updated](https://github.com/rancher/eks-operator/blob/1b8dfe2ea04eee43fd362c6da0e1cef2a09cb600/controller/eks-cluster-config-handler.go#L133) every time which reenqueues the handler immediately instead of doing a backoff reenqueue and is spamming the logs.

## Solution
 

My solution is to remove the metadata from the EKS logs

What happens now when an EKS error is logged?
- Raw err message is logged
- Remove metadata from the error. If removing metadata fails because the EKS error is not in the expected format, continue with the original logic to prevent a regression.  The new error string will contain the cluster name, error message, and node group.
- If the Status doesn't already contain the error (without metadata), add the error to the Status.

image

### CODE REVIEW

File: controller/eks-cluster-config-handler.go
Comment (cmurphy) [inline controller/eks-cluster-config-handler.go:122]: All the controllers have an `EnqueueAfter` method which allows you to reenqueue the handler after a time period:

https://github.com/rancher/eks-operator/blob/1b8dfe2ea04eee43fd362c6da0e1cef2a09cb600/controller/eks-cluster-config-handler.go#L61
https://github.com/rancher/eks-operator/blob/1b8dfe2ea04eee43fd362c6da0e1cef2a09cb600/controller/eks-cluster-config-handler.go#L256

Is there a way we could use that instead? Using a sleep here will just make the handler block waiting for the time to finish instead of returning control to the controller.

Comment (annablender) [reply controller/eks-cluster-config-handler.go:122]: Using `eksEnqueueAfter` still spams the logs every second. Using `Sleep` prevents the log spamming from what I can see

Comment (jakefhyde) [reply controller/eks-cluster-config-handler.go]: > LGTM, I also wonder if there is a better approach for solving this issue. Looking forward to hearing others' feedback.

I solved this at one point in encryption key rotation by setting an annotation with the time of the last sleep, and checking if the required interval (usually 1 second shorter than what we `EnqueueAfter` with) had passed since, and if it hadn't to not enqueue the object and return. There's also some fun stuff you can do with channels and tickers, but it's sort of niche and hard to get right.

**Comment by jakefhyde** [comment]
> LGTM, I also wonder if there is a better approach for solving this issue. Looking forward to hearing others' feedback.

I solved this at one point in encryption key rotation by setting an annotation with the time of the last sleep, and checking if the required interval (usually 1 second shorter than what we `EnqueueAfter` with) had passed since, and if it hadn't to not enqueue the object and return. There's also some fun stuff you can do with channels and tickers, but it's sort of niche and hard to get right.

File: controller/eks-cluster-config-handler.go
Comment (rmweir) [inline controller/eks-cluster-config-handler.go:122]: I don't think we want to block the thread like this. I think it's actually returning an error which reenqueues the controller since the status is a subresource iirc, so changes to it do cause a reenqueue. Reenqueueing on errors is the backboen to the framework and shouldn't be skipped like this. I would also think that we want the first occurrence to log immediately. There is a builtin backoff that should slow it down over time iirc. If you can show us an example of the offending error and we may be able to think of a different solution.

**Review by rmweir — CHANGES_REQUESTED** [review]

File: controller/eks-cluster-config-handler.go
Comment (jiaqiluo) [inline controller/eks-cluster-config-handler.go:127]: nit: simplify the logic 
```suggestion
			} 
			message = messageNoMeta
```

**Review by jiaqiluo — APPROVED** [review]

File: controller/eks-cluster-config-handler.go
Comment (annablender) [reply controller/eks-cluster-config-handler.go:126]: I believe my logic is correct here because we don't want the message to be set to `messageNoMeta` if the metadata removal returned an err

Comment (jiaqiluo) [reply controller/eks-cluster-config-handler.go:126]: I see. yeah, it makes sense. :)

**Review by jiaqiluo — APPROVED** [review]

File: controller/eks-cluster-config-handler.go
Comment (HarrisonWAffel) [inline controller/eks-cluster-config-handler.go:177]: `Strings.Index` will return `-1` if the substring isn't found, in which case `message[index:]` would panic. Will all error messages pass through this function?

Comment (rmweir) [reply controller/eks-cluster-config-handler.go:177]: Should check for -1 and return error if true.

**Review by rmweir — CHANGES_REQUESTED** [review]

File: controller/eks-cluster-config-handler.go
Comment (annablender) [reply controller/eks-cluster-config-handler.go:117]: Formatting. It looked really chunky on the page so I lengthened the comment to make it more readable and match the other lines

**Review by rmweir — DISMISSED** [review]
LGTM

**Review by rmweir — APPROVED** [review]

**Review by HarrisonWAffel — DISMISSED** [review]

**Review by jiaqiluo — APPROVED** [review]
Please squash two commits into one before merging. the 1st commit message is misleading now.

**Review by jiaqiluo — APPROVED** [review]

**Review by HarrisonWAffel — APPROVED** [review]

**Pull request merged**
